File: learnLeetCode/Day3/container_with_most_water.py
"""
Container With Most Water (Medium) — Bonus

Approach: Two pointers from both ends; move the shorter side inward.
Time: O(n), Space: O(1).
"""
from typing import List


def max_area(height: List[int]) -> int:
    left, right = 0, len(height) - 1
    best = 0
    while left < right:
        h = min(height[left], height[right])
        area = h * (right - left)
        if area > best:
            best = area
        if height[left] < height[right]:
            left += 1
        else:
            right -= 1
    return best


if __name__ == "__main__":
    assert max_area([1,8,6,2,5,4,8,3,7]) == 49
    assert max_area([1,1]) == 1
    assert max_area([1,2,1]) == 2
    print("container_with_most_water.py: all tests passed")

File: Day3/container_with_most_water.cpp
#include <iostream>
#include <vector>
using namespace std;

// Container With Most Water (Medium)
// Approach: Two pointers; move the shorter side inward. O(n) time, O(1) space.
int maxArea(const vector<int>& height) {
    int l = 0, r = (int)height.size() - 1;
    long long best = 0;
    while (l < r) {
        int h = min(height[l], height[r]);
        long long area = 1LL * h * (r - l);
        if (area > best) best = area;
        if (height[l] < height[r]) ++l; else --r;
    }
    return (int)best;
}

int main() {
    vector<int> h{1,8,6,2,5,4,8,3,7};
    cout << "max area: " << maxArea(h) << "\n";

    // basic checks
    if (maxArea({1,1}) != 1) return 1;
    if (maxArea({1,2,1}) != 2) return 1;
    return 0;
}
